crates/extensions/c8y_firmware_manager/src/actor.rs

```diff
@@ -0,0 +1,772 @@
+                "515" => match SmartRestFirmwareRequest::from_smartrest(&smartrest_message) {
+                    Ok(firmware_request) => {
+                        // Addressing a new firmware operation to further step.
+                        self.handle_firmware_download_request(firmware_request)
+                            .await
+                    }
+                    Err(_) => {
+                        error!("Incorrect c8y_Firmware SmartREST payload: {smartrest_message}");
+                        Ok(())
+                    }
+                },
+                _ => {
+                    // Ignore operation messages not meant for this plugin
+                    Ok(())
+                }
+            };
+
+            if let Err(err) = result {
+                error!("Handling of operation: '{smartrest_message}' failed with {err}");
+            }
+        }
+        Ok(())
+    }
+
+    // Validate the received SmartREST request and going to further step.
+    async fn handle_firmware_download_request(
+        &mut self,
+        smartrest_request: SmartRestFirmwareRequest,
+    ) -> Result<(), FirmwareManagementError> {
+        info!("Handling c8y_Firmware operation: {smartrest_request}");
+
+        if smartrest_request.device == self.config.tedge_device_id {
+            warn!("c8y-firmware-plugin does not support firmware operation for the main tedge device. \
+            Please define a custom operation handler for the c8y_Firmware operation.");
+            return Ok(());
+        }
+
+        let child_id = smartrest_request.device.as_str();
+
+        if let Err(err) = self
+            .validate_same_request_in_progress(smartrest_request.clone())
+            .await
+        {
+            return match err {
+                FirmwareManagementError::RequestAlreadyAddressed => {
+                    warn!("Skip the received c8y_Firmware operation as the same operation is already in progress.");
+                    Ok(())
+                }
+                _ => {
+                    self.fail_operation_in_cloud(child_id, None, &err.to_string())
+                        .await?;
+                    Err(err)
+                }
+            };
+        }
+
+        let op_id = nanoid!();
+        if let Err(err) =
+            // Addressing the new firmware operation to further step.
+            self
+                .handle_firmware_download_request_child_device(
+                    smartrest_request.clone(),
+                    op_id.as_str(),
+                )
+                .await
+        {
+            self.fail_operation_in_cloud(child_id, Some(&op_id), &err.to_string())
+                .await?;
+        }
+
+        Ok(())
+    }
+
+    // Check if the firmware file is already in cache.
+    // If yes, going to further step going to the last step; to publish a firmware request to child device.
+    // Otherwise, send a download request to the downloader.
```

Comment:
```suggestion
    // Check if the firmware file is already in cache.
    // If yes, publish a firmware request to child device with that firmware in the cache.
    // Otherwise, send a download request to the DownloaderActor and return immediately without waiting for the download to complete so that other requests/responses can be processed while the download is in progress.
    // The download will be performed by the DownloaderActor asynchronously and the response will be processed by this actor later on, in the `run` method.
```